File: packages/loupe-py/src/loupe/integrations/_providers.py
"""The canonical list of LLM providers Loupe auto-detects.

One source of truth for both the Python `loupe.integrations.httpx` patch and
the `loupe providers` CLI command. The TypeScript SDK mirrors this list in
`packages/loupe-ts/src/integrations/_providers.ts` — keep them in sync.

Entries are ordered by approximate 2026 popularity. The matcher walks the list
and returns the *first* hit, so put more-specific subdomains before broader
ones (e.g. `bedrock-runtime.*.amazonaws.com` before generic AWS domains).

Adding a provider:
1. Add an entry below.
2. Mirror it in the TS file.
3. Run `loupe providers` to confirm; run the tests.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Provider:
    label: str           # short canonical id
    host_suffix: str     # exact host or DNS suffix (without leading dot)
    name: str            # human label for `loupe providers`
    category: str        # frontier | inference | cloud | embedding | local | aggregator
    homepage: str = ""
    # Some providers (AWS regions, Vertex AI regions) embed an identifier in
    # the *middle* of the hostname rather than as a clean DNS suffix. For
    # those, use "contains" so we match `bedrock-runtime.us-east-1.amazonaws.com`
    # without also matching unrelated `lambda.amazonaws.com`.
    match_strategy: str = "suffix"  # "suffix" | "contains"


# ---------------------------------------------------------------------------
# Frontier labs (the model creators)
# ---------------------------------------------------------------------------

FRONTIER: list[Provider] = [
    Provider("anthropic", "api.anthropic.com", "Anthropic", "frontier", "https://anthropic.com"),
    Provider("openai", "api.openai.com", "OpenAI", "frontier", "https://openai.com"),
    Provider("gemini", "generativelanguage.googleapis.com", "Google Gemini", "frontier",
             "https://ai.google.dev"),
    Provider("mistral", "api.mistral.ai", "Mistral", "frontier", "https://mistral.ai"),
    Provider("mistral-codestral", "codestral.mistral.ai", "Mistral Codestral", "frontier"),
    Provider("cohere", "api.cohere.com", "Cohere", "frontier", "https://cohere.com"),
    Provider("cohere-legacy", "api.cohere.ai", "Cohere (legacy)", "frontier"),
    Provider("xai", "api.x.ai", "xAI Grok", "frontier", "https://x.ai"),
    Provider("deepseek", "api.deepseek.com", "DeepSeek", "frontier", "https://deepseek.com"),
    Provider("ai21", "api.ai21.com", "AI21 Labs", "frontier", "https://ai21.com"),
    Provider("reka", "api.reka.ai", "Reka", "frontier", "https://reka.ai"),
    Provider("aleph-alpha", "api.aleph-alpha.com", "Aleph Alpha", "frontier"),
    Provider("zhipu", "open.bigmodel.cn", "Zhipu GLM", "frontier"),
    Provider("baidu", "aip.baidubce.com", "Baidu ERNIE", "frontier"),
    Provider("alibaba", "dashscope.aliyuncs.com", "Alibaba Qwen", "frontier"),
]

# ---------------------------------------------------------------------------
# Inference providers (host open-weight models)
# ---------------------------------------------------------------------------

INFERENCE: list[Provider] = [
    Provider("groq", "api.groq.com", "Groq", "inference", "https://groq.com"),
    Provider("cerebras", "api.cerebras.ai", "Cerebras", "inference", "https://cerebras.ai"),
    Provider("sambanova", "api.sambanova.ai", "SambaNova", "inference", "https://sambanova.ai"),
    Provider("together", "api.together.xyz", "Together", "inference", "https://together.ai"),
    Provider("fireworks", "api.fireworks.ai", "Fireworks", "inference", "https://fireworks.ai"),
    Provider("deepinfra", "api.deepinfra.com", "DeepInfra", "inference"),
    Provider("hyperbolic", "api.hyperbolic.xyz", "Hyperbolic", "inference"),
    Provider("anyscale", "api.endpoints.anyscale.com", "Anyscale Endpoints", "inference"),
    Provider("nebius", "api.studio.nebius.ai", "Nebius AI Studio", "inference"),
    Provider("lambda", "api.lambdalabs.com", "Lambda Inference", "inference"),
    Provider("lepton", "api.lepton.ai", "Lepton AI", "inference"),
    Provider("siliconflow", "api.siliconflow.com", "SiliconFlow", "inference"),
    Provider("featherless", "api.featherless.ai", "Featherless", "inference"),
    Provider("inference-net", "api.inference.net", "Inference.net", "inference"),
    Provider("modal", "modal.com", "Modal Labs", "inference"),
    Provider("replicate", "api.replicate.com", "Replicate", "inference", "https://replicate.com"),
    Provider("perplexity", "api.perplexity.ai", "Perplexity", "inference", "https://perplexity.ai"),
]

# ---------------------------------------------------------------------------
# Aggregator gateways (route to many models)
# ---------------------------------------------------------------------------

AGGREGATORS: list[Provider] = [
    Provider("openrouter", "openrouter.ai", "OpenRouter", "aggregator"),
    Provider("portkey", "api.portkey.ai", "Portkey", "aggregator"),
    Provider("kong-ai", "ai-gateway.konghq.com", "Kong AI Gateway", "aggregator"),
    Provider("vellum", "api.vellum.ai", "Vellum", "aggregator"),
]

# ---------------------------------------------------------------------------
# Cloud-hosted (enterprise giants)
# ---------------------------------------------------------------------------

CLOUD: list[Provider] = [
    Provider("azure-openai", "openai.azure.com", "Azure OpenAI", "cloud"),
    Provider("aws-bedrock", "bedrock-runtime", "AWS Bedrock", "cloud",
             match_strategy="contains"),
    Provider("vertex-ai", "aiplatform.googleapis.com", "Google Vertex AI", "cloud",
             match_strategy="contains"),
    Provider("watsonx", "ml.cloud.ibm.com", "IBM watsonx", "cloud"),
    Provider("databricks", "cloud.databricks.com", "Databricks Model Serving", "cloud"),
]

# ---------------------------------------------------------------------------
# Embedding & retrieval providers
# ---------------------------------------------------------------------------

EMBEDDING: list[Provider] = [
    Provider("voyage", "api.voyageai.com", "Voyage AI", "embedding"),
    Provider("jina", "api.jina.ai", "Jina AI", "embedding"),
    Provider("nomic", "api-atlas.nomic.ai", "Nomic Atlas", "embedding"),
    Provider("huggingface", "api-inference.huggingface.co", "HuggingFace Inference", "embedding"),
    Provider("huggingface-endpoints", "endpoints.huggingface.cloud",
             "HuggingFace Endpoints", "embedding"),
]

# ---------------------------------------------------------------------------
# Local / self-hosted (servers users run on their own boxes)
# ---------------------------------------------------------------------------

LOCAL: list[Provider] = [
    Provider("local", "localhost", "Local server (Ollama / vLLM / LM Studio / LiteLLM)", "local"),
    Provider("local-ip", "127.0.0.1", "Local server (loopback)", "local"),
    Provider("local-net", "0.0.0.0", "Local server (any-iface)", "local"),
]


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

ALL_PROVIDERS: list[Provider] = (
    FRONTIER + INFERENCE + AGGREGATORS + CLOUD + EMBEDDING + LOCAL
)


def detect_provider_from_host(host: str | None) -> Provider | None:
    """Return the matching provider for a hostname, or None.

    For `match_strategy="suffix"` (the default), a hit means the host equals
    the suffix OR ends with `.{suffix}`. For `match_strategy="contains"`,
    the pattern just has to appear anywhere in the host — used for cloud
    providers where the identifier sits in the middle of the FQDN.

    Order matters — list-defined order is the matching order.
    """
    if not host:
        return None
    host = host.lower()
    for p in ALL_PROVIDERS:
        pat = p.host_suffix.lower()
        if p.match_strategy == "contains":
            if pat in host:
                return p
        elif host == pat or host.endswith("." + pat):
            return p
    return None


def looks_like_openai_compatible(body: object) -> bool:
    """Heuristic: does this request body look like an OpenAI-spec call?

    Catches LiteLLM-style proxies, internal gateways, OpenAI-compatible
    forks (Together, Fireworks et al. already speak OpenAI spec, but their
    URLs are known; this is for the *unknown* hosts).
    """
    if not isinstance(body, dict):
        return False
    return (
        "messages" in body
        and "model" in body
        and isinstance(body.get("messages"), list)
    )

File: packages/loupe-ts/src/integrations/_providers.ts
/**
 * The canonical list of LLM providers Loupe auto-detects. Mirror of the
 * Python `_providers.py`; keep them in sync.
 *
 * Entries are ordered by 2026 popularity. The matcher returns the *first*
 * hit — put more-specific subdomains before broader ones.
 */

export type ProviderCategory =
  | "frontier"
  | "inference"
  | "aggregator"
  | "cloud"
  | "embedding"
  | "local";

export interface Provider {
  label: string;
  hostSuffix: string;
  name: string;
  category: ProviderCategory;
  homepage?: string;
  /**
   * For providers that embed their identifier in the middle of the FQDN
   * (AWS regions, Vertex AI regions), set this to "contains".
   */
  matchStrategy?: "suffix" | "contains";
}

const FRONTIER: Provider[] = [
  { label: "anthropic", hostSuffix: "api.anthropic.com", name: "Anthropic", category: "frontier", homepage: "https://anthropic.com" },
  { label: "openai", hostSuffix: "api.openai.com", name: "OpenAI", category: "frontier", homepage: "https://openai.com" },
  { label: "gemini", hostSuffix: "generativelanguage.googleapis.com", name: "Google Gemini", category: "frontier", homepage: "https://ai.google.dev" },
  { label: "mistral", hostSuffix: "api.mistral.ai", name: "Mistral", category: "frontier", homepage: "https://mistral.ai" },
  { label: "mistral-codestral", hostSuffix: "codestral.mistral.ai", name: "Mistral Codestral", category: "frontier" },
  { label: "cohere", hostSuffix: "api.cohere.com", name: "Cohere", category: "frontier", homepage: "https://cohere.com" },
  { label: "cohere-legacy", hostSuffix: "api.cohere.ai", name: "Cohere (legacy)", category: "frontier" },
  { label: "xai", hostSuffix: "api.x.ai", name: "xAI Grok", category: "frontier", homepage: "https://x.ai" },
  { label: "deepseek", hostSuffix: "api.deepseek.com", name: "DeepSeek", category: "frontier", homepage: "https://deepseek.com" },
  { label: "ai21", hostSuffix: "api.ai21.com", name: "AI21 Labs", category: "frontier", homepage: "https://ai21.com" },
  { label: "reka", hostSuffix: "api.reka.ai", name: "Reka", category: "frontier", homepage: "https://reka.ai" },
  { label: "aleph-alpha", hostSuffix: "api.aleph-alpha.com", name: "Aleph Alpha", category: "frontier" },
  { label: "zhipu", hostSuffix: "open.bigmodel.cn", name: "Zhipu GLM", category: "frontier" },
  { label: "baidu", hostSuffix: "aip.baidubce.com", name: "Baidu ERNIE", category: "frontier" },
  { label: "alibaba", hostSuffix: "dashscope.aliyuncs.com", name: "Alibaba Qwen", category: "frontier" },
];

const INFERENCE: Provider[] = [
  { label: "groq", hostSuffix: "api.groq.com", name: "Groq", category: "inference", homepage: "https://groq.com" },
  { label: "cerebras", hostSuffix: "api.cerebras.ai", name: "Cerebras", category: "inference", homepage: "https://cerebras.ai" },
  { label: "sambanova", hostSuffix: "api.sambanova.ai", name: "SambaNova", category: "inference", homepage: "https://sambanova.ai" },
  { label: "together", hostSuffix: "api.together.xyz", name: "Together", category: "inference", homepage: "https://together.ai" },
  { label: "fireworks", hostSuffix: "api.fireworks.ai", name: "Fireworks", category: "inference", homepage: "https://fireworks.ai" },
  { label: "deepinfra", hostSuffix: "api.deepinfra.com", name: "DeepInfra", category: "inference" },
  { label: "hyperbolic", hostSuffix: "api.hyperbolic.xyz", name: "Hyperbolic", category: "inference" },
  { label: "anyscale", hostSuffix: "api.endpoints.anyscale.com", name: "Anyscale Endpoints", category: "inference" },
  { label: "nebius", hostSuffix: "api.studio.nebius.ai", name: "Nebius AI Studio", category: "inference" },
  { label: "lambda", hostSuffix: "api.lambdalabs.com", name: "Lambda Inference", category: "inference" },
  { label: "lepton", hostSuffix: "api.lepton.ai", name: "Lepton AI", category: "inference" },
  { label: "siliconflow", hostSuffix: "api.siliconflow.com", name: "SiliconFlow", category: "inference" },
  { label: "featherless", hostSuffix: "api.featherless.ai", name: "Featherless", category: "inference" },
  { label: "inference-net", hostSuffix: "api.inference.net", name: "Inference.net", category: "inference" },
  { label: "modal", hostSuffix: "modal.com", name: "Modal Labs", category: "inference" },
  { label: "replicate", hostSuffix: "api.replicate.com", name: "Replicate", category: "inference", homepage: "https://replicate.com" },
  { label: "perplexity", hostSuffix: "api.perplexity.ai", name: "Perplexity", category: "inference", homepage: "https://perplexity.ai" },
];

const AGGREGATORS: Provider[] = [
  { label: "openrouter", hostSuffix: "openrouter.ai", name: "OpenRouter", category: "aggregator" },
  { label: "portkey", hostSuffix: "api.portkey.ai", name: "Portkey", category: "aggregator" },
  { label: "kong-ai", hostSuffix: "ai-gateway.konghq.com", name: "Kong AI Gateway", category: "aggregator" },
  { label: "vellum", hostSuffix: "api.vellum.ai", name: "Vellum", category: "aggregator" },
];

const CLOUD: Provider[] = [
  { label: "azure-openai", hostSuffix: "openai.azure.com", name: "Azure OpenAI", category: "cloud" },
  { label: "aws-bedrock", hostSuffix: "bedrock-runtime", name: "AWS Bedrock", category: "cloud", matchStrategy: "contains" },
  { label: "vertex-ai", hostSuffix: "aiplatform.googleapis.com", name: "Google Vertex AI", category: "cloud", matchStrategy: "contains" },
  { label: "watsonx", hostSuffix: "ml.cloud.ibm.com", name: "IBM watsonx", category: "cloud" },
  { label: "databricks", hostSuffix: "cloud.databricks.com", name: "Databricks Model Serving", category: "cloud" },
];

const EMBEDDING: Provider[] = [
  { label: "voyage", hostSuffix: "api.voyageai.com", name: "Voyage AI", category: "embedding" },
  { label: "jina", hostSuffix: "api.jina.ai", name: "Jina AI", category: "embedding" },
  { label: "nomic", hostSuffix: "api-atlas.nomic.ai", name: "Nomic Atlas", category: "embedding" },
  { label: "huggingface", hostSuffix: "api-inference.huggingface.co", name: "HuggingFace Inference", category: "embedding" },
  { label: "huggingface-endpoints", hostSuffix: "endpoints.huggingface.cloud", name: "HuggingFace Endpoints", category: "embedding" },
];

const LOCAL: Provider[] = [
  { label: "local", hostSuffix: "localhost", name: "Local server (Ollama / vLLM / LM Studio / LiteLLM)", category: "local" },
  { label: "local-ip", hostSuffix: "127.0.0.1", name: "Local server (loopback)", category: "local" },
  { label: "local-net", hostSuffix: "0.0.0.0", name: "Local server (any-iface)", category: "local" },
];

export const ALL_PROVIDERS: Provider[] = [
  ...FRONTIER,
  ...INFERENCE,
  ...AGGREGATORS,
  ...CLOUD,
  ...EMBEDDING,
  ...LOCAL,
];

export function detectProviderFromHost(host: string | null | undefined): Provider | null {
  if (!host) return null;
  const h = host.toLowerCase();
  for (const p of ALL_PROVIDERS) {
    const pat = p.hostSuffix.toLowerCase();
    if (p.matchStrategy === "contains") {
      if (h.includes(pat)) return p;
    } else if (h === pat || h.endsWith(`.${pat}`)) {
      return p;
    }
  }
  return null;
}

export function looksLikeOpenAICompatible(body: unknown): boolean {
  if (!body || typeof body !== "object") return false;
  const b = body as Record<string, unknown>;
  return Array.isArray(b.messages) && "model" in b;
}
